decentralized-message-queue/src/DMQ/Protocol/LocalMsgNotification/Type.hs

```diff
@@ -0,0 +1,181 @@
+
+-- | todo haddock
+module DMQ.Protocol.LocalMsgNotification.Type where
+
+import Control.DeepSeq
+import Data.Kind
+import Data.List.NonEmpty
+import Network.TypedProtocol.Core
+import Ouroboros.Network.Util.ShowProxy
+
+-- | There are some constraints of the protocol that are not captured in the
+-- types of the messages, but are documented with the messages. Violation
+-- of these constraints is also a protocol error. The constraints are intended
+-- to ensure that implementations are able to work in bounded space.
+--
+instance Protocol (LocalMsgNotification msg) where
+  -- TODO haddock
+  -- | The messages in the protocol.
+  --
+  -- There are two ways to ask for messages, blocking and
+  -- non-blocking (todo: blocking only should suffice?).
+  -- They otherwise have the same semantics.
+  --
+  data Message (LocalMsgNotification msg) from to where
+
+    -- | Request a list of messages from the server,
+    --
+    --
+    -- * The blocking case is used when the server has announced it has no
+    --   further messages to provide, and otherwise the non-blocking request
+    --   must be used.
+    -- TODO: this request should have the max number of messages that shall
+    -- be returned?
+    MsgRequest
+      :: forall blocking msg.
+         SingBlockingStyle blocking
+      -> Message (LocalMsgNotification msg) StIdle (StBusy blocking)
+
+    -- | Reply with a list of messages, along with a flag indicating
+    -- whether the server has more messages that it can provide.
+    -- maybe this should just be always a blocking request?
+    --
+    MsgReply
+      :: forall blocking msg.
+         BlockingReplyList blocking msg
+      -> HasMore
+      -> Message (LocalMsgNotification msg) (StBusy blocking) StIdle
+
+    -- | Termination message, initiated by the server when the client is
+    -- making a blocking call for more messages.
+    --
+    MsgServerDone
+      :: Message (LocalMsgNotification msg) (StBusy StBlocking) StDone
+
+    MsgClientDone
+      :: Message (LocalMsgNotification msg) StIdle StDone
+
+  type StateAgency  (StBusy blocking)  = ServerAgency
+  type StateAgency  StIdle  = ClientAgency
+  type StateAgency  StDone  = NobodyAgency
+
+  type StateToken = SingMsgNotification
+
+instance ( NFData msg
+         ) => NFData (Message (LocalMsgNotification msg) from to) where
+  rnf (MsgRequest block)  = rnf block
+  rnf (MsgReply a !_more) = rnf a
+  rnf MsgServerDone{}     = ()
+  rnf MsgClientDone{}     = ()
+
+deriving instance (Eq msg) =>
+                  Eq (Message (LocalMsgNotification msg) from to)
+
+deriving instance (Show msg) =>
+                  Show (Message (LocalMsgNotification msg) from to)
+
+
+-- | The kind of the local message notification protocol, and the types of
+-- the states in the protocol state machine.
+--
+-- It is parameterised over the type of messages
+--
+type LocalMsgNotification :: Type -> Type
+type data LocalMsgNotification msg where
+  -- TODO haddock
+  -- |
```

Comment:
You can just remove the haddock placehoders and the TODO, the haddock attached to the type signature is enough.